### PR TITLE
fix(857): Support RHL proxied components in custom type checking

### PR DESCRIPTION
**Pull request by liviudobrea**
Closes carbon-design-system/carbon-components-react#857

This adds `react-hot-loader` as dependency, since new type checking is done via [areComponentsEqual](https://github.com/gaearon/react-hot-loader/blob/f3e1208878583136e9e14c0e2e6b7ffc7db94fea/src/utils.dev.js#L9-L11)
```javascript
export const areComponentsEqual = (a, b) =>
  getProxyOrType(a) === getProxyOrType(b)
``` 

#### Changelog

**New**

* react-hot-loader project dependency

**Changed**

* Updated childrenOf and childrenOfType type checking and tests

### CODE REVIEW

File: src/prop-types/childrenOf.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { Children } from 'react';
+import { areComponentsEqual } from 'react-hot-loader/patch';
```
Comment (asudoh) [inline src/prop-types/childrenOf.js]: @joshblack Would this add `react-hot-loader` dependency to our main React component library?

Comment (joshblack) [reply src/prop-types/childrenOf.js]: Looking at this, it seems so. Leaving a comment in the general thread.

**Comment by joshblack** [comment]
Hi @liviudobrea! 👋 How hard would it be to embed the `areComponentsEqual` check inside of the utility method versus bringing in RHL as a dependency?

**Comment by liviudobrea** [comment]
@joshblack That would be impossible, since the `areComponentsEqual` method uses an internal registry to check the type of the components you are requiring the equality check for. https://github.com/gaearon/react-hot-loader/blob/master/src/utils.dev.js and https://github.com/gaearon/react-hot-loader/blob/master/src/reconciler/proxies.js
The other method (that would not rely on having RHL as dep) is to check wether the components we're checking contains the `PROXY_KEY` prop https://github.com/gaearon/react-hot-loader/blob/df8ddaedbfd116e688ee7309a9b1f992318ebe41/src/proxy/constants.js#L1-L2.
```javascript
export const PREFIX = '__reactstandin__'
export const PROXY_KEY = `${PREFIX}key`
``` 
Personally, I would not wish for either (not for RHL as dependency, nor for checking for a propname that I'm uncertain of).

File: src/prop-types/childrenOf.js [hunk 2]
```diff
@@ -21,7 +22,13 @@ const childrenOf = expectedChildTypes => {
...
       const childDisplayName = getDisplayName(child.type || child);
-      if (!expectedChildTypes.includes(child.type)) {
+      const expectedChildType = expectedChildTypes.find(
+        c =>
+          typeof c === 'string'
+            ? c === childDisplayName
+            : c.name === childDisplayName
+      );
+      if (!areComponentsEqual(child.type, expectedChildType)) {
         throw new Error(
```
Comment (asudoh) [inline src/prop-types/childrenOf.js]: @joshblack Thinking more about this issue - Should the condition here *really* be considered fatal?

**Comment by joshblack** [comment]
@liviudobrea I'm down for checking to see if the component is being proxied. Is there a common way to access the underlying proxied component, or is this info stored in that registry that you alluded to?

File: src/prop-types/childrenOf.js (hunk 2 above)
Comment (joshblack) [reply src/prop-types/childrenOf.js]: Good point! Should be returning the error instead of throwing it 🙂 Must have missed this in an earlier PR.

**Comment by liviudobrea** [comment]
In RHL there's a helper method attached to every proxied component called
"__reactstandin__getCurrent()" that will return the original instance.

**Comment by joshblack** [comment]
@liviudobrea makes sense! Is it realistic to add a check that basically is looking to see if `__reactstandin__getCurrent` is defined, and if so use that to get access to the instance?

**Comment by asudoh** [comment]
@liviudobrea @joshblack As [this thread](https://github.com/gaearon/react-hot-loader/issues/304) discusses, I see a few possible solutions:

* Don’t treat non-matching React class fatal (Warning-only instead of error)
* `child.type` === `React.createElement(ExpectedComponentClass).type` (Or define a signature hash in the class and compare such signature)
* [RHL 4](https://github.com/gaearon/react-hot-loader/issues/304#issuecomment-353865002)?

**Comment by liviudobrea** [comment]
@asudoh 
- This would be preferred
- This would be unnecessarily expensive for our use case
- That link is broken. [This is the original reference](https://github.com/gaearon/react-hot-loader/#checking-element-types), and as you can see it makes use of `areComponentsEqual`, which requires `RHL` as project dependency.

@joshblack Until this is done https://github.com/gaearon/react-proxy/issues/68 (if ever...), I would currently opt for warning instead of throwing when comparing types, in favour of checking for the magic `__reactstandin__getCurrent` prop.

**Comment by joshblack** [comment]
Relevant PR in the Palantir codebase: https://github.com/palantir/blueprint/pull/2099

**Comment by joshblack** [comment]
Another option that (seemingly) could work: https://github.com/gatsbyjs/gatsby/issues/3486#issuecomment-364167698 

Maybe we can add it in behind a `__DEV__` flag?

**Comment by marijohannessen** [comment]
@liviudobrea @asudoh @joshblack we'll be closing this PR in 7 days if there is no more activity by then!

**Comment by asudoh** [comment]
@liviudobrea My apologies in case you were waiting for my feedback. One of the links @joshblack referred to uses the 2nd approach. My preference is the 1st approach, unless RHL has solved the problem - Hope it makes things clear. Thanks!

**Comment by marijohannessen** [comment]
Closing due to inactivity. Feel free to re-open if needed.

**Pull request closed without merging**